File: lib/weaver.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import { chatJSON } from "./providers/openai";
import { neighborIds } from "./graph";
import { z } from "zod";
import type {
  GraphEdgeRow,
  GraphNodeRow,
  MemoryRow,
  ProvenanceRow,
  Relative,
} from "./types";

export type GapType = "missing_origin" | "orphan_object" | "unrelated_person";

export interface Gap {
  type: GapType;
  nodeId: string;
}

export interface WeaverData {
  nodes: GraphNodeRow[];
  edges: GraphEdgeRow[];
  provenance: ProvenanceRow[];
  memories: Pick<MemoryRow, "id" | "contributor_id" | "kind" | "summary">[];
  /** person node ids that have at least one enrolled face */
  facePersonIds: string[];
  wearerNodeId: string | null;
  relatives: Relative[];
  openQuestionRelativeIds: string[];
}

const GAP_ORDER: GapType[] = [
  "missing_origin",
  "orphan_object",
  "unrelated_person",
];
const ORIGIN_RELS = new Set(["origin", "started_by", "taught_by"]);

function edgesTouching(nodeId: string, edges: GraphEdgeRow[]) {
  return edges.filter(
    (e) => e.from_node === nodeId || e.to_node === nodeId
  );
}

/** distinct contributors whose memories provide provenance for a node or its edges */
function contributorsOf(d: WeaverData, nodeId: string): Set<string> {
  const edgeIds = new Set(edgesTouching(nodeId, d.edges).map((e) => e.id));
  const out = new Set<string>();
  for (const p of d.provenance) {
    if (p.node_id === nodeId || (p.edge_id && edgeIds.has(p.edge_id))) {
      out.add(p.contributor_id);
    }
  }
  return out;
}

function memoryIdsTouchingNode(d: WeaverData, nodeId: string): Set<string> {
  const edgeIds = new Set(edgesTouchingNodeIds(d, nodeId).map((e) => e.id));
  const out = new Set<string>();
  for (const p of d.provenance) {
    if (p.node_id === nodeId || (p.edge_id && edgeIds.has(p.edge_id))) {
      out.add(p.memory_id);
    }
  }
  return out;
}

function edgesTouchingNodeIds(d: WeaverData, nodeId: string) {
  return edgesTouching(nodeId, d.edges);
}

/** Rule-based gap detection. Types are checked in a fixed order. */
export function findGaps(d: WeaverData): Gap[] {
  const gaps: Gap[] = [];
  const byId = new Map(d.nodes.map((n) => [n.id, n]));

  for (const n of d.nodes) {
    if (n.type !== "tradition" && n.type !== "event") continue;
    const incoming = d.edges.filter(
      (e) => e.to_node === n.id && e.rel === "participates_in"
    ).length;
    const hasOrigin = edgesTouching(n.id, d.edges).some((e) =>
      ORIGIN_RELS.has(e.rel)
    );
    if (incoming >= 2 && !hasOrigin) {
      gaps.push({ type: "missing_origin", nodeId: n.id });
    }
  }

  for (const n of d.nodes) {
    if (n.type !== "object") continue;
    const memoryCount = new Set(
      d.provenance.filter((p) => p.node_id === n.id).map((p) => p.memory_id)
    ).size;
    if (memoryCount !== 1) continue;
    const linked = edgesTouching(n.id, d.edges).some((e) => {
      const other = byId.get(e.from_node === n.id ? e.to_node : e.from_node);
      return (
        other && ["tradition", "event", "place"].includes(other.type)
      );
    });
    if (!linked) gaps.push({ type: "orphan_object", nodeId: n.id });
  }

  for (const n of d.nodes) {
    if (n.type !== "person") continue;
    if (!d.facePersonIds.includes(n.id)) continue;
    if (n.relation_to_wearer) continue;
    const touchesWearer =
      d.wearerNodeId &&
      d.edges.some(
        (e) =>
          (e.from_node === n.id && e.to_node === d.wearerNodeId) ||
          (e.to_node === n.id && e.from_node === d.wearerNodeId)
      );
    if (!touchesWearer) gaps.push({ type: "unrelated_person", nodeId: n.id });
  }

  return gaps;
}

/** degree(node) + 2 * distinct contributors providing provenance for it */
export function gapScore(d: WeaverData, gap: Gap): number {
  const degree = edgesTouching(gap.nodeId, d.edges).length;
  return degree + 2 * contributorsOf(d, gap.nodeId).size;
}

export function pickTopGap(d: WeaverData): Gap | null {
  const gaps = findGaps(d);
  if (!gaps.length) return null;
  return gaps.sort((a, b) => {
    const diff = gapScore(d, b) - gapScore(d, a);
    if (diff !== 0) return diff;
    return GAP_ORDER.indexOf(a.type) - GAP_ORDER.indexOf(b.type);
  })[0];
}

/**
 * Who to ask. Relatives whose memories already describe the gap node itself
 * are skipped: the point is to ask someone whose contributions sit NEXT to
 * the gap. Ties go to a relative who uploaded a related photo.
 */
export function routeQuestion(d: WeaverData, gap: Gap): string | null {
  const neighbors = neighborIds(gap.nodeId, d.edges);
  const describers = contributorsOf(d, gap.nodeId);
  const open = new Set(d.openQuestionRelativeIds);

  let candidates = d.relatives.filter(
    (r) => !open.has(r.id) && !describers.has(r.id)
  );
  if (!candidates.length) {
    candidates = d.relatives.filter((r) => !open.has(r.id));
  }
  if (!candidates.length) return null;

  const memoriesByContributor = new Map<string, typeof d.memories>();
  for (const m of d.memories) {
    const list = memoriesByContributor.get(m.contributor_id) ?? [];
    list.push(m);
    memoriesByContributor.set(m.contributor_id, list);
  }

  const neighborEdgeIds = new Set<string>();
  for (const nid of neighbors) {
    edgesTouching(nid, d.edges).forEach((e) => neighborEdgeIds.add(e.id));
  }
  const touchesNeighbor = (memoryId: string) =>
    d.provenance.some(
      (p) =>
        p.memory_id === memoryId &&
        ((p.node_id && neighbors.has(p.node_id)) ||
          (p.edge_id && neighborEdgeIds.has(p.edge_id)))
    );
  const touchesGap = (memoryId: string) =>
    memoryIdsTouchingNode(d, gap.nodeId).has(memoryId);

  const scored = candidates.map((r) => {
    const mems = memoriesByContributor.get(r.id) ?? [];
    const neighborMems = mems.filter((m) => touchesNeighbor(m.id));
    const gapMems = mems.filter((m) => touchesGap(m.id));
    const score = 2 * neighborMems.length + gapMems.length;
    const hasRelatedPhoto = [...neighborMems, ...gapMems].some(
      (m) => m.kind === "photo"
    );
    return { relative: r, score, hasRelatedPhoto };
  });

  scored.sort((a, b) => {
    if (b.score !== a.score) return b.score - a.score;
    if (a.hasRelatedPhoto !== b.hasRelatedPhoto) return a.hasRelatedPhoto ? -1 : 1;
    return 0;
  });
  return scored[0]?.relative.id ?? null;
}

const questionZod = z.object({ question: z.string() });

/** Full DB-backed Weaver run: detect gap, pick target, phrase, insert. */
export async function runWeaver(
  sb: SupabaseClient,
  familyId: string
): Promise<{ question: Record<string, unknown> | null; gap: Gap | null }> {
  const [nodes, edges, prov, mems, faces, relatives, questions, wearer] =
    await Promise.all([
      sb.from("graph_nodes").select("*").eq("family_id", familyId),
      sb.from("graph_edges").select("*").eq("family_id", familyId),
      sb.from("provenance").select("*"),
      sb.from("memories").select("id, contributor_id, kind, summary").eq("family_id", familyId),
      sb.from("face_embeddings").select("person_node_id").eq("family_id", familyId),
      sb.from("relatives").select("*").eq("family_id", familyId),
      sb.from("weaver_questions").select("target_relative_id").eq("family_id", familyId).eq("status", "open"),
      sb.from("graph_nodes").select("id").eq("family_id", familyId).eq("relation_to_wearer", "self").limit(1),
    ]);

  const memoryRows = (mems.data ?? []) as WeaverData["memories"];
  const memFamily = new Set(memoryRows.map((m) => m.id));
  const data: WeaverData = {
    nodes: (nodes.data ?? []) as GraphNodeRow[],
    edges: (edges.data ?? []) as GraphEdgeRow[],
    provenance: ((prov.data ?? []) as ProvenanceRow[]).filter((p) =>
      memFamily.has(p.memory_id)
    ),
    memories: memoryRows,
    facePersonIds: (faces.data ?? []).map((f) => f.person_node_id),
    wearerNodeId: wearer.data?.[0]?.id ?? null,
    relatives: (relatives.data ?? []) as Relative[],
    openQuestionRelativeIds: (questions.data ?? []).map(
      (q) => q.target_relative_id
    ),
  };

  const gap = pickTopGap(data);
  if (!gap) return { question: null, gap: null };
  const targetId = routeQuestion(data, gap);
  if (!targetId) return { question: null, gap };

  const gapNode = data.nodes.find((n) => n.id === gap.nodeId)!;
  const evidenceIds = [...memoryIdsTouchingNode(data, gap.nodeId)].slice(0, 3);
  const evidence = evidenceIds
    .map((id) => {
      const m = data.memories.find((x) => x.id === id);
      return m
        ? { memory_id: m.id, contributor_id: m.contributor_id, summary: m.summary }
        : null;
    })
    .filter(Boolean) as { memory_id: string; contributor_id: string; summary: string }[];
  // Neighbor memories make good evidence too when the gap itself is thin.
  if (evidence.length < 2) {
    const neighbors = neighborIds(gap.nodeId, data.edges);
    for (const m of data.memories) {
      if (evidence.length >= 3) break;
      if (evidence.some((e) => e.memory_id === m.id)) continue;
      const touches = data.provenance.some(
        (p) => p.memory_id === m.id && p.node_id && neighbors.has(p.node_id)
      );
      if (touches) {
        evidence.push({
          memory_id: m.id,
          contributor_id: m.contributor_id,
          summary: m.summary,
        });
      }
    }
  }

  const nameOf = (id: string) =>
    data.relatives.find((r) => r.id === id)?.name ?? "Someone";
  const evidenceText = evidence
    .map((e) => `${nameOf(e.contributor_id)}: "${e.summary}"`)
    .join("\n");

  let questionText: string;
  try {
    const res = await chatJSON<{ question: string }>({
      name: "weaver_question",
      jsonSchema: {
        type: "object",
        additionalProperties: false,
        required: ["question"],
        properties: { question: { type: "string" } },
      },
      zodSchema: questionZod,
      system:
        "You are the Kin Family Weaver. You ask one relative a warm, low-pressure question to fill a gap in the family memory. One or two sentences. Cite who said what. End with a specific question. No pressure, no guilt.",
      user: `Gap type: ${gap.type}\nNode: "${gapNode.label}" (${gapNode.type})\nEvidence:\n${evidenceText || "(no direct evidence)"}\n\nWrite the question.`,
    });
    questionText = res.question;
  } catch {
    questionText = `The family remembers "${gapNode.label}" but no one has recorded where it came from. Do you remember?`;
  }

  const { data: row, error } = await sb
    .from("weaver_questions")
    .insert({
      family_id: familyId,
      target_relative_id: targetId,
      gap_node_id: gap.nodeId,
      gap_type: gap.type,
      question_text: questionText,
      evidence,
    })
    .select()
    .single();
  if (error) throw error;
  return { question: row, gap };
}

File: brain/src/weaver.rs
//! The Family Weaver: finds holes in the family's memory graph and decides
//! which relative is most likely to be able to fill each one.
//!
//! Port of the pure half of lib/weaver.ts. Behavior is pinned by
//! conformance/weaver-fixtures.json.

use std::collections::{BTreeSet, HashMap, HashSet};

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GapType {
    MissingOrigin,
    OrphanObject,
    UnrelatedPerson,
}

/// Detection order, and the tie-break when two gaps score equally.
const GAP_ORDER: [GapType; 3] =
    [GapType::MissingOrigin, GapType::OrphanObject, GapType::UnrelatedPerson];

/// Edge relations that answer "where did this come from".
const ORIGIN_RELS: [&str; 3] = ["origin", "started_by", "taught_by"];

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Gap {
    #[serde(rename = "type")]
    pub gap_type: GapType,
    #[serde(rename = "nodeId")]
    pub node_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphNode {
    pub id: String,
    #[serde(rename = "type")]
    pub node_type: String,
    pub label: String,
    pub relation_to_wearer: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphEdge {
    pub id: String,
    pub from_node: String,
    pub rel: String,
    pub to_node: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Provenance {
    pub id: String,
    pub memory_id: String,
    pub node_id: Option<String>,
    pub edge_id: Option<String>,
    pub contributor_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WeaverMemory {
    pub id: String,
    pub contributor_id: String,
    pub kind: String,
    pub summary: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Relative {
    pub id: String,
    pub name: String,
    pub relation_to_wearer: String,
    pub color: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WeaverData {
    pub nodes: Vec<GraphNode>,
    pub edges: Vec<GraphEdge>,
    pub provenance: Vec<Provenance>,
    pub memories: Vec<WeaverMemory>,
    /// Person node ids with at least one enrolled face.
    #[serde(rename = "facePersonIds")]
    pub face_person_ids: Vec<String>,
    #[serde(rename = "wearerNodeId")]
    pub wearer_node_id: Option<String>,
    pub relatives: Vec<Relative>,
    #[serde(rename = "openQuestionRelativeIds")]
    pub open_question_relative_ids: Vec<String>,
}

fn edges_touching<'a>(node_id: &str, edges: &'a [GraphEdge]) -> Vec<&'a GraphEdge> {
    edges
        .iter()
        .filter(|e| e.from_node == node_id || e.to_node == node_id)
        .collect()
}

fn neighbor_ids(node_id: &str, edges: &[GraphEdge]) -> HashSet<String> {
    let mut out = HashSet::new();
    for e in edges {
        if e.from_node == node_id {
            out.insert(e.to_node.clone());
        }
        if e.to_node == node_id {
            out.insert(e.from_node.clone());
        }
    }
    out
}

/// Distinct contributors whose memories provide provenance for a node or its edges.
fn contributors_of(d: &WeaverData, node_id: &str) -> BTreeSet<String> {
    let edge_ids: HashSet<&str> =
        edges_touching(node_id, &d.edges).iter().map(|e| e.id.as_str()).collect();
    d.provenance
        .iter()
        .filter(|p| {
            p.node_id.as_deref() == Some(node_id)
                || p.edge_id.as_deref().is_some_and(|eid| edge_ids.contains(eid))
        })
        .map(|p| p.contributor_id.clone())
        .collect()
}

fn memory_ids_touching_node(d: &WeaverData, node_id: &str) -> HashSet<String> {
    let edge_ids: HashSet<&str> =
        edges_touching(node_id, &d.edges).iter().map(|e| e.id.as_str()).collect();
    d.provenance
        .iter()
        .filter(|p| {
            p.node_id.as_deref() == Some(node_id)
                || p.edge_id.as_deref().is_some_and(|eid| edge_ids.contains(eid))
        })
        .map(|p| p.memory_id.clone())
        .collect()
}

/// Rule-based gap detection. Types are checked in a fixed order, which is
/// also the order the results come back in.
pub fn find_gaps(d: &WeaverData) -> Vec<Gap> {
    let mut gaps = Vec::new();
    let by_id: HashMap<&str, &GraphNode> = d.nodes.iter().map(|n| (n.id.as_str(), n)).collect();

    // A shared tradition or event that nobody has explained the origin of.
    for n in &d.nodes {
        if n.node_type != "tradition" && n.node_type != "event" {
            continue;
        }
        let incoming = d
            .edges
            .iter()
            .filter(|e| e.to_node == n.id && e.rel == "participates_in")
            .count();
        let has_origin = edges_touching(&n.id, &d.edges)
            .iter()
            .any(|e| ORIGIN_RELS.contains(&e.rel.as_str()));
        if incoming >= 2 && !has_origin {
            gaps.push(Gap { gap_type: GapType::MissingOrigin, node_id: n.id.clone() });
        }
    }

    // An object mentioned exactly once and tied to no occasion.
    for n in &d.nodes {
        if n.node_type != "object" {
            continue;
        }
        let memory_count: HashSet<&str> = d
            .provenance
            .iter()
            .filter(|p| p.node_id.as_deref() == Some(n.id.as_str()))
            .map(|p| p.memory_id.as_str())
            .collect();
        if memory_count.len() != 1 {
            continue;
        }
        let linked = edges_touching(&n.id, &d.edges).iter().any(|e| {
            let other_id = if e.from_node == n.id { &e.to_node } else { &e.from_node };
            by_id.get(other_id.as_str()).is_some_and(|o| {
                matches!(o.node_type.as_str(), "tradition" | "event" | "place")
            })
        });
        if !linked {
            gaps.push(Gap { gap_type: GapType::OrphanObject, node_id: n.id.clone() });
        }
    }

    // A face the family enrolled but nobody placed relative to the wearer.
    for n in &d.nodes {
        if n.node_type != "person" || !d.face_person_ids.contains(&n.id) {
            continue;
        }
        if n.relation_to_wearer.is_some() {
            continue;
        }
        let touches_wearer = d.wearer_node_id.as_deref().is_some_and(|w| {
            d.edges.iter().any(|e| {
                (e.from_node == n.id && e.to_node == w) || (e.to_node == n.id && e.from_node == w)
            })
        });
        if !touches_wearer {
            gaps.push(Gap { gap_type: GapType::UnrelatedPerson, node_id: n.id.clone() });
        }
    }

    gaps
}

/// degree(node) + 2 * distinct contributors providing provenance for it.
pub fn gap_score(d: &WeaverData, gap: &Gap) -> usize {
    edges_touching(&gap.node_id, &d.edges).len() + 2 * contributors_of(d, &gap.node_id).len()
}

pub fn pick_top_gap(d: &WeaverData) -> Option<Gap> {
    let mut gaps = find_gaps(d);
    if gaps.is_empty() {
        return None;
    }
    let order = |t: GapType| GAP_ORDER.iter().position(|g| *g == t).unwrap_or(usize::MAX);
    // Stable ID breaks ties independently of database row order.
    gaps.sort_by(|a, b| {
        gap_score(d, b)
            .cmp(&gap_score(d, a))
            .then_with(|| order(a.gap_type).cmp(&order(b.gap_type)))
            .then_with(|| a.node_id.cmp(&b.node_id))
    });
    gaps.into_iter().next()
}

/// Who to ask. Relatives whose memories already describe the gap node are
/// skipped: the point is to ask someone whose contributions sit NEXT to the
/// gap, not someone who already told us what little we know. Ties go to a
/// relative who uploaded a related photo.
pub fn route_question(d: &WeaverData, gap: &Gap) -> Option<String> {
    let neighbors = neighbor_ids(&gap.node_id, &d.edges);
    let touching = memory_ids_touching_node(d, &gap.node_id);
    let describers: HashSet<String> = d.memories.iter().filter(|m| m.kind != "photo" && touching.contains(&m.id)).map(|m| m.contributor_id.clone()).collect();
    let open: HashSet<&str> = d.open_question_relative_ids.iter().map(String::as_str).collect();

    let mut candidates: Vec<&Relative> = d
        .relatives
        .iter()
        .filter(|r| !open.contains(r.id.as_str()) && !describers.contains(&r.id))
        .collect();
    if candidates.is_empty() {
        candidates = d.relatives.iter().filter(|r| !open.contains(r.id.as_str())).collect();
    }
    if candidates.is_empty() {
        return None;
    }

    let mut neighbor_edge_ids: HashSet<&str> = HashSet::new();
    for nid in &neighbors {
        for e in edges_touching(nid, &d.edges) {
            neighbor_edge_ids.insert(e.id.as_str());
        }
    }
    let touches_neighbor = |memory_id: &str| -> bool {
        d.provenance.iter().any(|p| {
            p.memory_id == memory_id
                && (p.node_id.as_deref().is_some_and(|n| neighbors.contains(n))
                    || p.edge_id.as_deref().is_some_and(|e| neighbor_edge_ids.contains(e)))
        })
    };
    let gap_memories = memory_ids_touching_node(d, &gap.node_id);

    struct Scored<'a> {
        relative: &'a Relative,
        score: usize,
        has_related_photo: bool,
    }

    let mut scored: Vec<Scored> = candidates
        .into_iter()
        .map(|r| {
            let mems: Vec<&WeaverMemory> =
                d.memories.iter().filter(|m| m.contributor_id == r.id).collect();
            let neighbor_mems: Vec<&WeaverMemory> =
                mems.iter().copied().filter(|m| touches_neighbor(&m.id)).collect();
            let gap_mems: Vec<&WeaverMemory> =
                mems.iter().copied().filter(|m| gap_memories.contains(&m.id)).collect();
            let has_related_photo = neighbor_mems
                .iter()
                .chain(gap_mems.iter())
                .any(|m| m.kind == "photo");
            Scored {
                relative: r,
                score: 2 * neighbor_mems.len() + gap_mems.len(),
                has_related_photo,
            }
        })
        .collect();

    // Match the TypeScript stable-ID tie break.
    scored.sort_by(|a, b| {
        b.score
            .cmp(&a.score)
            .then_with(|| b.has_related_photo.cmp(&a.has_related_photo))
            .then_with(|| a.relative.id.cmp(&b.relative.id))
    });
    scored.first().map(|s| s.relative.id.clone())
}
